src/main/kotlin/me/zeroeightsix/kami/feature/module/Notebot.kt

```diff
@@ -0,0 +1,213 @@
+            }
+        }
+    }
+    private fun loadSong() {
+        mc.player?.let {
+            playingSong = false
+            MidiParser.parse("songs${File.separator}$songName").let {
+                noteSequence = it.first
+                channelMap = it.second
+            }
+            if (noteSequence.isEmpty())
+                snackbarMessage(it, "Unable to find MIDI: songs${File.separator}$songName")
+            else
+                snackbarMessage(it, "Loaded song ${File.separator}$songName")
+        }
+        lastNote = System.currentTimeMillis()
+        elapsed = 0
+    }
+
+    @EventHandler
+    val listener = Listener<TickEvent.InGame>({
+        if (playingSong) {
+            val player = it.player
+            val world = it.world
+            if (!player.isCreative) {
+                while (noteSequence.isNotEmpty() && noteSequence.firstKey() <= elapsed) {
+                    playNotes(noteSequence.pollFirstEntry().value, player, world)
+                }
+                if (noteSequence.isEmpty()) playingSong = false
+                elapsed += System.currentTimeMillis() - lastNote
+            } else snackbarMessage(player, "You are in creative mode and cannot play music.")
+        }
+        lastNote = System.currentTimeMillis()
+    })
+
+    private fun playNotes(notes: List<Note>, player: ClientPlayerEntity, world: ClientWorld) {
+        val blockPosArr: ArrayList<BlockPos?> = ArrayList()
+        notes.forEach { n ->
+            val channelsArray: Array<Instrument> =
+                arrayOf(channelZero, channelOne, channelTwo, channelThree, channelFour)
+            val number = (n.track % (channelsArray.size - 1))
+            val enum = channelsArray[if (number < 0) 0 else number]
+            snackbarMessage(player, enum.toString())
+            blockPosArr.add(instrumentMap[enum][if (enum == Instrument.HAT && hatAlwaysAsFSharp) 0 else n.notebotNote])
+        }
+        playBlock(blockPosArr, player, world)
+    }
+
+    private fun playBlock(blockPosTracks: ArrayList<BlockPos?>, player: ClientPlayerEntity, world: ClientWorld) {
+        if (blockPosTracks.isNotEmpty()) {
+            blockPosTracks.forEach { blockPos ->
+                if (blockPos != null) {
+                    getIrreplaceableNeighbour(world, blockPos)?.let { (_, direction) ->
+                        mc.networkHandler?.sendPacket(
+                            PlayerActionC2SPacket(PlayerActionC2SPacket.Action.START_DESTROY_BLOCK, blockPos, direction)
+                        )
+                        mc.networkHandler?.sendPacket(
+                            PlayerActionC2SPacket(PlayerActionC2SPacket.Action.ABORT_DESTROY_BLOCK, blockPos, direction)
+                        )
+                        player.swingHand(Hand.MAIN_HAND)
+
+                        mc.setCooldown(4)
+                    }
+                }
+            }
+        }
+    }
+
+    fun snackbarMessage(player: ClientPlayerEntity, t: String) {
```

Comment:
Yarn mappings call this an `actionBar` and not a `snackbar`, consider renaming this method. (See ClientPlayerEntity#sendMessage)